### PR TITLE
Make Slab::new const on Rust 1.39+, Use #[track_caller] on Rust 1.46+

### PR DESCRIPTION
**Pull request by taiki-e**
Closes #87
Closes #90

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -225,8 +225,25 @@ impl<T> Slab<T> {
     /// # use slab::*;
     /// let slab: Slab<i32> = Slab::new();
     /// ```
-    pub fn new() -> Slab<T> {
-        Slab::with_capacity(0)
+    #[cfg(not(slab_no_const_vec_new))]
+    pub const fn new() -> Self {
```
Comment (Darksonn) [inline src/lib.rs:229]: The fact that const depends on the version should probably be documented.

**Review by Darksonn — APPROVED** [review]

File: src/lib.rs (hunk 1 above)
Comment (notgull) [inline src/lib.rs:229]: You could use the [`const-fn`](https://docs.rs/const_fn/latest/const_fn/) crate here to avoid code deduplication.

Comment (taiki-e) [reply src/lib.rs:231]: Updated docs.

Comment (taiki-e) [reply src/lib.rs:231]: It is my crate, but it doesn't yet implement the behavior I want.

https://github.com/tokio-rs/slab/blob/56e794722ade3f07ecd928d7f66933290e36ac1d/build.rs#L15-L17

**Pull request merged**